zipkin-collector/rabbitmq/src/main/java/zipkin/collector/rabbitmq/RabbitMqCollector.java

```diff
@@ -0,0 +1,213 @@
+import zipkin.collector.CollectorMetrics;
+import zipkin.collector.CollectorSampler;
+import zipkin.internal.LazyCloseable;
+import zipkin.storage.StorageComponent;
+
+import static zipkin.internal.Util.checkNotNull;
+
+/**
+ * This collector consumes encoded binary messages from a RabbitMQ queue.
+ */
+public final class RabbitMqCollector implements CollectorComponent {
+  private static final Logger LOG = LoggerFactory.getLogger(RabbitMqCollector.class);
+
+  public static Builder builder() {
+    return new Builder();
+  }
+
+  /** Configuration including defaults needed to consume spans from a RabbitMQ queue. */
+  public static final class Builder implements CollectorComponent.Builder {
+    Collector.Builder delegate = Collector.builder(RabbitMqCollector.class);
+    CollectorMetrics metrics = CollectorMetrics.NOOP_METRICS;
+    String queue = "zipkin";
+    ConnectionFactory connectionFactory;
+    List<String> addresses;
+    int concurrency = 1;
+
+    @Override public Builder storage(StorageComponent storage) {
+      this.delegate.storage(storage);
+      return this;
+    }
+
+    @Override public Builder sampler(CollectorSampler sampler) {
+      this.delegate.sampler(sampler);
+      return this;
+    }
+
+    @Override public Builder metrics(CollectorMetrics metrics) {
+      this.metrics = checkNotNull(metrics, "metrics").forTransport("rabbitmq");
+      this.delegate.metrics(this.metrics);
+      return this;
+    }
+
+    public Builder addresses(List<String> addresses) {
+      this.addresses = addresses;
+      return this;
+    }
+
+    public Builder concurrency(int concurrency) {
+      this.concurrency = concurrency;
+      return this;
+    }
+
+    public Builder connectionFactory(ConnectionFactory connectionFactory) {
+      this.connectionFactory = checkNotNull(connectionFactory, "connectionFactory");
+      return this;
+    }
+
+    /**
+     * Queue zipkin spans will be consumed from. Defaults to "zipkin-spans".
+     */
+    public Builder queue(String queue) {
+      this.queue = checkNotNull(queue, "queue");
+      return this;
+    }
+
+    @Override public RabbitMqCollector build() {
+      return new RabbitMqCollector(this);
+    }
+  }
+
+  private final LazyRabbitWorkers rabbitWorkers;
+
+  RabbitMqCollector(Builder builder) {
+    this.rabbitWorkers = new LazyRabbitWorkers(builder);
+  }
+
+  @Override
+  public RabbitMqCollector start() {
+    this.rabbitWorkers.get();
+    return this;
+  }
+
+  @Override
+  public CheckResult check() {
+    try {
+      CheckResult failure = this.rabbitWorkers.failure.get();
+      if (failure != null) return failure;
+      return CheckResult.OK;
+    } catch (RuntimeException e) {
+      return CheckResult.failed(e);
+    }
+  }
+
+  @Override
+  public void close() throws IOException {
+    this.rabbitWorkers.close();
+  }
+
+  static final class LazyRabbitWorkers extends LazyCloseable<ExecutorService> {
+
+    final Builder builder;
```

Comment:
careful about saving builder here as it is mutable and start is different thread than .build(). one way is to make your builder auto-value and make a copy of the builder (or just save off the params). otherwise add a TODO mentioning the edge-case mutability concern